File: tutorials/anchor-browser-agent/data-collection-example.js
const { chromium } = require('playwright-core');
const axios = require('axios');

const ANCHOR_API_KEY = process.env.ANCHOR_API_KEY;

if (!ANCHOR_API_KEY) {
    console.error("❌ ANCHOR_API_KEY is not set. Run:");
    console.error("export ANCHOR_API_KEY=<your-api-key>");
    process.exit(1);
}
async function createSession(ANCHOR_API_KEY) {
    const url = "https://api.anchorbrowser.io/v1/sessions";
    const payload = {
        session: {
            timeout: {
                max_duration: 4,
                idle_timeout: 2
            }
        }
    };
    const headers = {
        "anchor-api-key": ANCHOR_API_KEY,
        "Content-Type": "application/json"
    };

    try {
        // Initialize session
        const response = await axios.post(url, payload, { headers });
        return response.data;
    } catch (error) {
        if (error.response) {
            const { status, statusText, data } = error.response;
            if (status === 401) {
                console.error("❌ Authentication failed: Invalid or missing API key");
            } else {
                console.error(`❌ HTTP Error ${status}: ${statusText}`);
                console.error("Response data:", data);
            }
        } else if (error.request) {
            console.error("❌ Network error: No response from server");
        } else {
            console.error("❌ Error:", error.message);
        }
        process.exit(1);
    }
}

async function dataCollectionExample(connectionString) {
    const browser = await chromium.connectOverCDP(connectionString);
    const context = browser.contexts()[0];
    const ai = context.serviceWorkers()[0];
    const page = context.pages()[0];

    await page.goto(
        "https://play.grafana.org/a/grafana-k8s-app/navigation/nodes?from=now-1h&to=now&refresh=1m",
        { waitUntil: 'domcontentloaded' }
    );
    const result = await ai.evaluate('Collect the node names and their CPU average %, return in JSON array');
    await browser.close();
    return result;
}

(async () => {
    try {
        const sessionData = await createSession(ANCHOR_API_KEY);
        console.log(sessionData);
        const connectionString = sessionData.data.cdp_url;
        const result = await dataCollectionExample(connectionString);
        console.log(result);
        process.exit(0);
    } catch (error) {
        console.error("❌ Runtime error:", error.message);
        console.log("Please check the data-collection guide for more troubleshooting.");
        process.exit(1);
    }
})();

File: tutorials/anchor-browser-agent/data-collection-example.py
import os
import requests
from playwright.sync_api import sync_playwright

ANCHOR_API_KEY = os.getenv("ANCHOR_API_KEY")

# Create a session
def create_session(ANCHOR_API_KEY):
    url = "https://api.anchorbrowser.io/v1/sessions"

    payload = {
        "session": {
            "timeout": {                # Reduced timeouts for cost-effective learning
                "max_duration": 4,
                "idle_timeout": 2
            }
        }
    }
    headers = {
        "anchor-api-key": ANCHOR_API_KEY,
        "Content-Type": "application/json"
    }

    # Initialize session
    response = requests.request("POST", url, json=payload, headers=headers)

    # Make sure to keep the cdpUrl for the next step
    session_data = response.json()
    return session_data

def data_collection_example(connection_string):
    with sync_playwright() as p:
        browser = p.chromium.connect_over_cdp(connection_string)
        context = browser.contexts[0]
        target_url = "chrome-extension://bppehibnhionalpjigdjdilknbljaeai/background.js"
        ai = next((sw for sw in context.service_workers if sw.url == target_url), None)
        page = context.pages[0]

        page.goto(
            "https://play.grafana.org/a/grafana-k8s-app/navigation/nodes?from=now-1h&to=now&refresh=1m",
            wait_until="domcontentloaded"
        )
        result = ai.evaluate('Collect the node names and their CPU average %, return in JSON array')
        return result

if not ANCHOR_API_KEY:
    raise Exception("ANCHOR_API_KEY is not set, please run `export ANCHOR_API_KEY=<your-api-key>` and try again.")

session_data = create_session(ANCHOR_API_KEY)
print(session_data)
connection_string = session_data['data']['cdp_url']
result = data_collection_example(connection_string)
print(result)



